docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst

```diff
@@ -0,0 +1,421 @@
+        sink.asExpr() = call.getAnArgument()
+      )
+    }
+  }
+
+  module Flow = DataFlow::Global<Config>;
+
+  from Flow::PathNode source, Flow::PathNode sink
+  where Flow::flowPath(source, sink)
+  select sink.getNode(), source, sink, "Flow from user input to sink!"
+
+will catch most things á la
+
+.. code-block:: cpp
+
+  void fill_structure(B* b, const int* pu) {
+    // ...
+    b->a->p = pu;
+  }
+
+  void process_structure(const B* b) {
+    sink(*b->a->p);
+  }
+
+  void get_and_process() {
+    int u = user_input();
+    B* b = (B*)malloc(sizeof(B));
+    // ...
+    fill_structure(b, &u);
+    // ...
+    process_structure(b);
+    free(b);
+  }
+
+This is simple to match because we see:
+  1. User input starts at ``user_input()`` and flows into ``fill_structure``.
+  2. The data is written to the object ``b`` with access path ``[a, p]``.
+  3. The object ``b`` flows out of ``fill_structure`` and into ``process_structure``.
+  4. The access path ``[a, p]`` is read in ``process_structure`` and the value ends up in the sink.
+
+Flow from a qualifier to a field access
+---------------------------------------
+
+However, sometimes the writes or reads are not visible to CodeQL (for example, because the implementation of the function isn’t included in the database), and so dataflow won't be able to match up all stores with reads, and thus you don't get the result you want. For example, consider the following example:
+
+.. code-block:: cpp
+
+  void write_user_input_to(void*);
+  void use_value(int);
+  void* malloc(unsigned long);
+
+  struct U {
+    const int* p;
+    int x;
+  };
+
+  void process_user_data(const int* p) {
+    // ...
+    use_value(*p);
+  }
+
+  void get_and_process_user_input_v2() {
+    U* u = (U*)malloc(sizeof(U));
+    write_user_input_to(u);
+    process_user_data(u->p);
+    free(u);
+  }
+
+Here, flow starts at the outgoing argument of ``write_user_input_to(...)`` and proceeds to ``u->p``. However, because CodeQL has not observed a write to p prior to the read ``u->p``, dataflow will stop at ``u``. In order to convince CodeQL to proceed we need to add an additional flow step through field reads like so:
+
+.. code-block:: ql
+
+  /**
+  * @kind path-problem
+  */
+
+  import semmle.code.cpp.dataflow.new.DataFlow
+  import Flow::PathGraph
+
+  module Config implements DataFlow::ConfigSig {
+    predicate isSource(DataFlow::Node source) {
+      exists(Call call |
+        call.getTarget().hasName("write_user_input_to") and
+        source.asDefiningArgument() = call.getArgument(0)
+      )
+    }
+
+    predicate isSink(DataFlow::Node sink) {
+      exists(Call call |
+        call.getTarget().hasName("use_value") and
+        sink.asExpr() = call.getAnArgument()
+      )
+    }
+
+    predicate isAdditionalFlowStep(DataFlow::Node n1, DataFlow::Node n2) {
+      exists(FieldAccess fa |
+        n1.asIndirectExpr() = fa.getQualifier() and
+        n2.asIndirectExpr() = fa
+      )
```

Comment:
I've added a small comment about this in 0a71705dfe5f2fc73d37e1f05a2fe9ae99504ccd. Let me know if you think we should actually do the optimization you suggested instead of simply mentioning it in the prose.

Reply:
The added comment is definitely good.  I would be tempted to implement it in the code as well (because people love to copy-paste code from tutorials), but I can see that there's a design decision involved in whether you're matching a type name or field or whatever so it isn't one size fits all.

Reply:
I don't have strong opinions about this though.  It's fine as it is (now).